### PR TITLE
Introduce job tree struct

### PR DESCRIPTION
**Pull request by orlandohohmeier**
---
ℹ️   _This branch depends on changes that will be introduced with #250, thus it will fail until the respective branch is merged._

---

Introduce basic job tree struct – this struct will be used to type and abstract the Chronos API response. It provide the needed functionality to parse and create sub trees (if needed), to insert jobs in the correct location based on id/path matching.

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @mlunoe

File: src/js/structs/JobTree.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+module.exports = class JobTree extends Tree {
+  /**
+   * (Chronos) JobTree
+   * @param {{
+   *          id:string,
+   *          items:array<({id:string, items:array}|*)>,
+   *          filterProperties:{propertyName:(null|string|function)}
+   *        }} options
+   * TODO (orlandohohmeier): Add remove item method
```
Comment (mlunoe) [inline src/js/structs/JobTree.js]: Let's add the JIRA task number here

File: src/js/structs/JobTree.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    super({filterProperties});
+
+    this.id = '/';
+
+    if (typeof id == 'string') {
+      this.id = id;
+    }
+
+    // Parse and add items to the current tree
+    if (Util.isArray(items)) {
```
Comment (mlunoe) [inline src/js/structs/JobTree.js]: We should remove this Util. We are using `Array.isArray` now

File: src/js/structs/JobTree.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+   * @param {JobTree|Job} item
+   */
+  add(item) {
+    if (!(item instanceof Job || item instanceof JobTree)) {
+      throw new TypeError('item is neither an instance of Job nor JobTree');
+    }
+
+    if (!item.getId().startsWith(this.getId())) {
+      throw new Error(
+        `item id (${item.getId()})  doesn't match tree id (${this.getId()})`
```
Comment (mlunoe) [inline src/js/structs/JobTree.js]: I think you have an extra space here in between `)` and `doesn't`

File: src/js/structs/JobTree.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+   * Add item - This method will create sub trees if needed to insert the item
+   * at the correct location (based on id/path matching).
+   * @param {JobTree|Job} item
+   */
+  add(item) {
+    if (!(item instanceof Job || item instanceof JobTree)) {
+      throw new TypeError('item is neither an instance of Job nor JobTree');
+    }
+
+    if (!item.getId().startsWith(this.getId())) {
```
Comment (pierlo-upitup) [inline src/js/structs/JobTree.js]: How about caching `item.getId()` so we can reuse it in the next 3 occurences?

File: src/js/structs/JobTree.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Check if the item is already present in the tree (including sub trees)
+    if (this.findItemById(item.getId())) {
+      // TODO: DCOS-7370 - Update item (type Job) data
+      return;
+    }
+
+    // Get the parent id (e.g. /group) by matching every thing but the item
+    // name including the preceding slash "/" (e.g. /id).
+    let [parentId] = item.getId().match(/\/.*?(?=\/?[^/]+\/?$)/);
```
Comment (pierlo-upitup) [inline src/js/structs/JobTree.js]: are we protected against `.getId()` returning a string with no leading `/`? for example `group/sub/test` here would become `/sub`

File: src/js/structs/JobTree.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Add item to the current tree if it's the actual parent tree
+    if (this.getId() === parentId) {
+      super.add(item);
```
Comment (pierlo-upitup) [inline src/js/structs/JobTree.js]: ❓

File: src/js/structs/__tests__/JobTree-test.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+          new JobTree({id: '/group/foo', items: []}),
+          {id: '/group/bar', items: []},
+          {id: '/group/alpha'},
+          new Job({id: '/group/beta'}),
+          {id: '/group/wibble/wobble'}
+        ],
+      });
+    });
+
+    it('defaults id to root tree (groups) id', function () {
```
Comment (pierlo-upitup) [inline src/js/structs/__tests__/JobTree-test.js]: Let's try to have human readable descriptions 😄 . How about `it(creates the root group if no items are supplied)`?

File: src/js/structs/__tests__/JobTree-test.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      let tree = new JobTree({items: []});
+      expect(tree.getId()).toEqual('/');
+    });
+
+    it('sets correct tree (groups) id', function () {
```
Comment (pierlo-upitup) [inline src/js/structs/__tests__/JobTree-test.js]: how about `it(creates the group ids correctly)` and since we're at it how about adding 
```
expect(this.instance.getId()).toEqual('/group');
expect(this.instance.getItems()[0].getId()).toEqual('/group/foo');
```
?

Comment (pierlo-upitup) [inline src/js/structs/__tests__/JobTree-test.js]: NVM – I see you perform similar checks later.

**Comment by pierlo-upitup** [comment]
A few comments and failing tests needs fixing. 
```
 Runtime Error
Error: Cannot find module '../Job' from 'JobTree-test.js'
    at Runtime._resolveNodeModule (/Users/pierlo/Development/dcos-ui/node_modules/jest-cli/src/Runtime/Runtime.js:451:11)
    at Object.<anonymous> (/Users/pierlo/Development/dcos-ui/src/js/structs/__tests__/JobTree-test.js:1:169)
    at jasmine2 (/Users/pierlo/Development/dcos-ui/node_modules/jest-jasmine2/src/index.js:231:16)
    at process._tickCallback (node.js:406:9)
```

**Comment by orlandohohmeier** [comment]
@pierlo-upitup The test fail at the moment as this depends on stuff that will be introduced with #250.

File: src/js/structs/__tests__/JobTree-test.js (hunk 8 above)
Comment (orlandohohmeier) [inline src/js/structs/__tests__/JobTree-test.js]: This test verifies that `JobTree` is using the default (root tree) group id, if no id is provided.

File: src/js/structs/JobTree.js (hunk 5 above)
Comment (orlandohohmeier) [inline src/js/structs/JobTree.js]: Good catch! We will probably never ever get such data for the API, but I'll harden our implementation to make sure we're not creating wrong structures based on corrupted data.

File: src/js/structs/JobTree.js (hunk 4 above)
Comment (orlandohohmeier) [inline src/js/structs/JobTree.js]: Yeah, should make it a little easier to read the lines.

File: src/js/structs/JobTree.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    if (this.findItemById(item.getId())) {
+      // TODO: DCOS-7370 - Update item (type Job) data
+      return;
+    }
+
+    // Get the parent id (e.g. /group) by matching every thing but the item
+    // name including the preceding slash "/" (e.g. /id).
+    let [parentId] = item.getId().match(/\/.*?(?=\/?[^/]+\/?$)/);
+
+    // Add item to the current tree if it's the actual parent tree
```
Comment (orlandohohmeier) [inline src/js/structs/JobTree.js]: We're trying to find the parent of an item (`Job` or `JobTree`), by extracting the`parentId` from the `itemId`. We're adding it to the current tree if the `parentId` matches the current `JobTree` id, otherwise we're trying to find the parent (`findItemById`) or create a new parent (`JobTree`) and add the item to that one.
**TLDR;** This method recursively creates all the parents needed to insert an item at the correct spot (based on it's id).

File: src/js/structs/JobTree.js (hunk 6 above)
Comment (orlandohohmeier) [inline src/js/structs/JobTree.js]: This call adds an item to the current list of items.

File: src/js/structs/JobTree.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+          'and should not end with a slash, except for root id which is only ' +
+          'a slash.');
+      }
+
+      this.id = id;
+    }
+
+    // Parse and add items to the current tree
+    if (Array.isArray(items)) {
+      items.forEach((item)=> {
```
Comment (mlunoe) [inline src/js/structs/JobTree.js]: #nit missing space after `)` and before `=>`
